synapseclient/core/download/download_functions.py

```diff
@@ -0,0 +1,727 @@
+    elif cached_file_path is not None:
+        # file already cached so use that as the download location
+        download_location = os.path.dirname(cached_file_path)
+    else:
+        # file not cached and no user-specified location so default to .synapseCache
+        download_location = synapse_cache_location
+
+    # resolve file path collisions by either overwriting, renaming, or not downloading, depending on the
+    # ifcollision value
+    download_path = resolve_download_path_collisions(
+        download_location=download_location,
+        file_name=file_name,
+        if_collision=if_collision,
+        synapse_cache_location=synapse_cache_location,
+        cached_file_path=cached_file_path,
+    )
+    if download_path is None:
+        return
+
+    if cached_file_path is not None:  # copy from cache
+        if download_path != cached_file_path:
+            # create the foider if it does not exist already
+            if not os.path.exists(download_location):
+                os.makedirs(download_location)
+            shutil.copy(cached_file_path, download_path)
+
+    else:  # download the file from URL (could be a local file)
+        object_type = "FileEntity" if submission is None else "SubmissionAttachment"
+        object_id = entity["id"] if submission is None else submission
+
+        # reassign downloadPath because if url points to local file (e.g. file://~/someLocalFile.txt)
+        # it won't be "downloaded" and, instead, downloadPath will just point to '~/someLocalFile.txt'
+        # _downloadFileHandle may also return None to indicate that the download failed
+        download_path = await download_by_file_handle(
+            file_handle_id=entity.dataFileHandleId,
+            synapse_id=object_id,
+            entity_type=object_type,
+            destination=download_path,
+        )
+
+        if download_path is None or not os.path.exists(download_path):
+            return
+
+    # converts the path format from forward slashes back to backward slashes on Windows
+    entity.path = os.path.normpath(download_path)
+    entity.files = [os.path.basename(download_path)]
+    entity.cacheDir = os.path.dirname(download_path)
+
+
+async def download_by_file_handle(
+    file_handle_id: str,
+    synapse_id: str,
+    entity_type: str,
+    destination: str,
+    # TODO: Update this retries to be time based to match the upload logic
+    retries: int = 5,
+    synapse_client: Optional["Synapse"] = None,
+) -> str:
+    """
+    Download a file from the given URL to the local file system.
+
+    Arguments:
+        file_handle_id: The id of the FileHandle to download
+        synapse_id:     The id of the Synapse object that uses the FileHandle e.g. "syn123"
+        entity_type:   The type of the Synapse object that uses the FileHandle e.g. "FileEntity"
+        destination:  The destination on local file system
+        retries:      The Number of download retries attempted before throwing an exception.
```

Comment:
nit: Add `Default is 5.`

Reply:
I'll be removing this shortly to convert this to be a time based retry system.